cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -189,8 +189,25 @@ inline std::optional<AtomKey> get_key_for_version_query(
     const std::shared_ptr<VersionMapEntry>& version_map_entry, 
     const pipelines::VersionQuery& version_query) {
     return util::variant_match(version_query.content_,
-        [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) {
-            return find_index_key_for_version_id(specific_version.version_id_, version_map_entry);
+        [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) -> std::optional<AtomKey> {
+            auto signed_version_id = specific_version.version_id_;
+            VersionId version_id;
+            if (signed_version_id >= 0) {
+                version_id = static_cast<VersionId>(signed_version_id);
+            } else {
+                auto opt_latest = version_map_entry->get_first_index(true);
+                if (opt_latest.has_value()) {
+                    auto opt_version_id = get_version_id_negative_index(opt_latest->version_id(), signed_version_id);
```

Comment:
Looking at the implementation of `get_version_id_negative_index`, no special treatment exist if `signed_version_id` is negative and strictly smaller than `static_cast<SignedVersionId>(latest) - 1`, (a negative value for will be returned via `opt_version_id`).

Should ArcticDB adopt python's list's behavior, i.e.:
```python
>>> a = [1, 2, 3]
>>> a[-2]  
2
>>> a[-7]  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```

and err similarly in this case?

Reply:
This is exactly how it behaves?
`// Given the latest version, and a negative index into the version map, returns the desired version ID or std::nullopt if it would be negative
inline std::optional<VersionId> get_version_id_negative_index(VersionId latest, SignedVersionId index) {
    internal::check<ErrorCode::E_ASSERTION_FAILURE>(index < 0, "get_version_id_negative_index expects a negative index, received {}", index);
    // +1 so that as_of=-1 returns the latest entry
    auto candidate_version_id = static_cast<SignedVersionId>(latest) + index + 1;
    return candidate_version_id >= 0 ? std::make_optional<VersionId>(static_cast<VersionId>(candidate_version_id)) : std::nullopt;
}`
If `candidate_version_id` is negative, it returns `std::nullopt`, which further up the stack either throws an exception, or returns a `DataError` object

Reply:
I think that my remark was not clear: a `std::nullopt` is indeed returned in this case, eventually returning a `DataError` to users.

As of my comprehension of the implementation, the error will not be explicit and it might be hard for users to know for which reason a `DataError` is returned. Currently, `DataError` are documented as being returned when the version is missing:

https://github.com/man-group/ArcticDB/blob/664b7597c249f46cc2c0e14646046b6df7e3aff9/python/arcticdb/version_store/library.py#L877-L882

Anyway that's a UX suggestion, we could address later in another PX.

Reply:
Ah, I understand now. This behaviour is consistent with positive `as_of` (i.e. the error is the same if there are only 3 versions, but you request version 4). You're right that we don't currently have a good way to distinguish between the symbol doesn't exist, the symbol exists but the version has been deleted, the symbol exists but the version never existed, and the symbol exists but the requested version number cannot exist (i.e. is negative).